x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/PhaseCacheManagement.java

```
@@ -157,15 +156,19 @@ public static boolean updateIndicesForPolicy(
             .filter(meta -> isIndexPhaseDefinitionUpdatable(xContentRegistry, client, meta, newPolicy.getPolicy(), licenseState))
             .toList();
 
-        final List<String> refreshedIndices = new ArrayList<>(indicesThatCanBeUpdated.size());
+        final List<IndexMetadata.Builder> refreshedIndices = new ArrayList<>(indicesThatCanBeUpdated.size());
         for (IndexMetadata index : indicesThatCanBeUpdated) {
             try {
-                refreshPhaseDefinition(projectMetadataBuilder, index, newPolicy);
-                refreshedIndices.add(index.getIndex().getName());
+                var idxBuilder = prepareRefreshPhaseDefinition(index, newPolicy);
+                refreshedIndices.add(idxBuilder);
             } catch (Exception e) {
                 logger.warn(() -> format("[%s] unable to refresh phase definition for updated policy [%s]", index, newPolicy.getName()), e);
             }
```

I am probably missing something. Since we catch exception with a warning log here and continue with the next index, wouldn't exception down the stack still result in partial update, seems we just delay putting the partial index metadata into the project metadata? I wonder if we should throw an exception instead?

---

The new logic batches up the list of changed index metadata objects and only adds them to the project metadata builder after collecting all of them. Overall, it's a very small refactoring, and practically speaking shouldn't modify any behavior. The bigger picture here is that I'm trying to make changes to the method contract to ensure that the project metadata builder passed is less likely to be modified unnecessarily since it may be carrying changes from another batch of cluster state operations.

---

I am very much fine with the small refactoring (without behavior change). I probably misunderstood the PR description `This saves from polluting a potentially shared project metadata builder instance with incomplete state updates in the event that an exception is thrown from lower down the call stack` thinking it means  "all or nothing" refresh on indices. But it just means that we don't update the project metadata until all indices are processed (with or without exception).